### PR TITLE
feat(deploykey): Add deploykey resource

### LINKED ISSUE #13: Implement Deploy keys managed resource
### What problem are you facing?
provider-gitlab does not have support for gitlab [deploy keys](https://docs.gitlab.com/ee/user/project/deploy_keys/) resource

### How could Crossplane help solve your problem?
Implement https://docs.gitlab.com/ee/api/deploy_keys.html resource.

### PR DESCRIPTION
**Pull request by lvvorovi**
### Description of your changes

Fixes #13 

I have:

- [ ] Read and followed Crossplane's [contribution process].
- [ ] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: apis/projects/v1alpha1/deploykey_type.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	ProjectIdRef *xpv1.Reference `json:"projectIdRef,omitempty"`
+
+	// ProjectIDSelector selects reference to a project to retrieve its projectId.
+	// +optional
+	// +immutable
+	ProjectIdSelector *xpv1.Selector `json:"projectIdSelector,omitempty"`
+
+	// New Deploy Key’s title
+	// required
+	Title string `json:"title,omitempty"`
```
Comment (MisterMX) [inline apis/projects/v1alpha1/deploykey_type.go:45]: Required fields don't need `omitempty`
```suggestion
	Title string `json:"title"`
```

File: apis/projects/v1alpha1/deploykey_type.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	// +immutable
+	ProjectIdRef *xpv1.Reference `json:"projectIdRef,omitempty"`
+
+	// ProjectIDSelector selects reference to a project to retrieve its projectId.
+	// +optional
+	// +immutable
+	ProjectIdSelector *xpv1.Selector `json:"projectIdSelector,omitempty"`
+
+	// New Deploy Key’s title
+	// required
```
Comment (MisterMX) [inline apis/projects/v1alpha1/deploykey_type.go:44]:
```suggestion
	// New Deploy Key’s title.
	// This property is required.
```

File: apis/projects/v1alpha1/deploykey_type.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	CanPush *bool `json:"canPush,omitempty"`
+
+	// Expiration date for the Deploy Key. Does not expire if no value is provided.
+	// Expected in ISO 8601 format (2019-03-15T08:00:00Z)
+	// +optional
+	ExpiresAt *metav1.Time `json:"expiresAt,omitempty"`
+
+	// KeySecretRef field representing referrence to the key
+	// +required
+	KeySecretRef xpv1.SecretKeySelector `json:"keySecretRef,omitempty"`
```
Comment (MisterMX) [inline apis/projects/v1alpha1/deploykey_type.go:58]: To mark a field as required it is enough to omit the `omitempty`. `+required` is not necessary.

```suggestion
	KeySecretRef xpv1.SecretKeySelector `json:"keySecretRef"`
```

File: apis/projects/v1alpha1/referencers.go [hunk 4]
```diff
@@ -119,6 +119,29 @@ func (mg *DeployToken) ResolveReferences(ctx context.Context, c client.Reader) e
 	return nil
 }
 
+// ResolveReferences of this Deploy Key
+func (mg *DeployKey) ResolveReferences(ctx context.Context, c client.Reader) error {
```
Comment (MisterMX) [inline apis/projects/v1alpha1/referencers.go:123]: To save you time and complexity I recommend to use `angryjet` to generate the `ResolveReferences` for your types, like here: https://github.com/crossplane-contrib/provider-aws/blob/master/apis/eks/manualv1alpha1/nodegroup_types.go#L65

Generation is already included in `make generate`.

File: examples/groups/group.yaml [hunk 5]
```diff
@@ -15,4 +15,4 @@ spec:
   # a reference to a Kubernetes secret to which the controller will write the runnersToken
   writeConnectionSecretToRef:
     name: gitlab-group-example-group
-    namespace: crossplane-system
+    namespace: crossplane-system
```
Comment (MisterMX) [inline examples/groups/group.yaml:18]: Keep LF at the end of files

File: examples/projects/deploykey.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    keySecretRef:
+      namespace: <example-name-space>
+      name: <example-name>
+      key: <example-key>
+  providerConfigRef:
+    name: <gitlab-provider>
+  # a reference to a Kubernetes secret to which the controller will write the runnersToken
+  writeConnectionSecretToRef:
+    name: gitlab-example-deploy-key
+    namespace: crossplane-system
```
Comment (MisterMX) [inline examples/projects/deploykey.yaml:20]: Keep LF at the end of files

File: examples/projects/project.yaml [hunk 7]
```diff
@@ -14,4 +14,4 @@ spec:
   # a reference to a Kubernetes secret to which the controller will write the runnersToken
   writeConnectionSecretToRef:
     name: gitlab-project-example-project
-    namespace: crossplane-system
+    namespace: crossplane-system
```
Comment (MisterMX) [inline examples/projects/project.yaml:17]: Keep LF at the end of files

File: examples/providerconfig/provider.yaml [hunk 8]
```diff
@@ -11,4 +12,4 @@ spec:
     secretRef:
       namespace: crossplane-system
       name: gitlab-credentials
-      key: token
+      key: token
```
Comment (MisterMX) [inline examples/providerconfig/provider.yaml:15]: Keep LF at the end of files

File: pkg/controller/projects/deploykeys/controller.go [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+const (
+	ERR_NOT_DEPLOY_KEY = "managed resource is not a Gitlab deploy key custom resource"
+	ERR_NOT_FOUDN      = "404 project deploy key not found"
+	ERR_GET_FAIL       = "cannot get Gitlab deploy key"
+	ERR_CREATE_FAIL    = "cannot create Gitlab deploy key"
+	ERR_UPDATE_FAIL    = "cannot update Gitlab deploy key"
+	ERR_DELETE_FAIL    = "cannot delete Gitlab deploy key"
+	ERR_KEY_MISSING    = "missing key ref value"
+	ERR_ID_NOT_INT     = "external-name is not an int"
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:37]: Unlike in other languages, constants or any other kind of variable should not have UPPER_CASE letters. Instead it should be like `errNotDeployKey`.

Btw: Alle Identifiers starting with an upper case letter are considered public. All with a lower case letter are private (package) scope.

File: pkg/controller/projects/deploykeys/controller.go [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+)
+
+var (
+	deployKeyDoesNotExist = managed.ExternalObservation{}
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:41]: This field is not necessary and should be inlined instead.

File: pkg/controller/projects/deploykeys/controller.go [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+	}
+
+	return &external{kube: deployKeyService.kube, client: deployKeyService.newGitlabClientFn(*config)}, nil
+}
+
+func (deployKeyService *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	customResource, ok := mg.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return deployKeyDoesNotExist, errors.New(ERR_NOT_DEPLOY_KEY)
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:83]: Please just use plain `managed.ExternalObservation{}` for simplicity. It initializes an empty struct with all its fields set to their default values (`false`, `nil`).

```suggestion
		return managed.ExternalObservation{}, errors.New(ERR_NOT_DEPLOY_KEY)
```

File: pkg/controller/projects/deploykeys/controller.go [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+	isLateInitialized := !compare.Equal(currentState, &customResource.Spec.ForProvider)
+
+	customResource.Status.AtProvider = v1alpha1.DeployKeyObservation{}
+
+	customResource.Status.SetConditions(xpv1.Available())
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        true,
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:118]: Following this logic, the controller will never call `Update` since you always set `ResourceUpToDate` to `true`. You should implement a function `isUpToDate` that compares `cr` with `deployKey` and checks if any of the updatable fields has changed.

File: pkg/controller/projects/deploykeys/controller.go [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+
+	id := strconv.Itoa(keyResponse.ID)
+
+	meta.SetExternalName(customResource, id)
+
+	connectionDetails := managed.ConnectionDetails{} //Anything to be added?
+
+	return managed.ExternalCreation{
+		ExternalNameAssigned: true,
+		ConnectionDetails:    connectionDetails,
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:164]: You don't have to specify `ConnectionDetails` if there is nothing to expose.

File: pkg/controller/projects/deploykeys/controller.go [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+func (deployKeyService *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	customResource, ok := mg.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(ERR_NOT_DEPLOY_KEY)
+	}
+
+	_, _, err := deployKeyService.client.UpdateDeployKey(
+		customResource.Spec.ForProvider.ProjectId,
+		*customResource.Status.AtProvider.Id,
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:177]: Always use `meta.GetExternalName`.

File: pkg/controller/projects/deploykeys/controller_test.go [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+	}
+
+	deployKey.Spec.ForProvider.ProjectId = &testProjectId
+
+	return deployKey
+}
+
+func TestObserve(t *testing.T) {
+	type expected struct {
+		customResource resource.Managed
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller_test.go:113]: I would use `cr` or `deployKey` and type `DeployKey` directly.

**Review by MisterMX — CHANGES_REQUESTED** [review]
@lvvorovi thank you for you contribution. This looks already pretty good. Please look at the remarks I have made. Also, maybe you could use less empty lines to seperate code lines as it makes them hard to read.

File: pkg/controller/projects/deploykeys/controller.go [hunk 16]
```diff
@@ -263,3 +257,14 @@ func generateUpdateOptions(customResourse *v1alpha1.DeployKey) *gitlab.UpdateDep
 		CanPush: customResourse.Spec.ForProvider.CanPush,
 	}
 }
+
+func isUpToDate(cr *v1alpha1.DeployKey, dk *gitlab.ProjectDeployKey) bool {
+	isCanPushUpToDate := *cr.Spec.ForProvider.CanPush == dk.CanPush
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:262]: cr.Spec.ForProvider.CanPush is a pointer and might be `nil`.

You can use the deref functions from [`k8s.io/util/pointer`](https://pkg.go.dev/k8s.io/utils/pointer):

```suggestion
	isCanPushUpToDate := pointer.BoolEqual(cr.Spec.ForProvider.CanPush, &dk.CanPush)
```

File: pkg/controller/projects/deploykeys/controller.go [hunk 17]
```diff
@@ -263,3 +257,14 @@ func generateUpdateOptions(customResourse *v1alpha1.DeployKey) *gitlab.UpdateDep
...
+	isTitleUpToDate := cr.Spec.ForProvider.Title == dk.Title
+
+	if isCanPushUpToDate && isTitleUpToDate {
+		return true
+	}
+
+	return false
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:269]:
```suggestion
	return isCanPushUpToDate && isTitleUpToDate
```

**Review by MisterMX — CHANGES_REQUESTED** [review]

**Review by MisterMX — CHANGES_REQUESTED** [review]

File: examples/groups/group.yaml
[the remark this answers is not in the document]
Comment (lvvorovi) [reply examples/groups/group.yaml:1]: I added a new file, guess that was the reason. No changes. Pulled original file from master.

File: apis/projects/v1alpha1/deploykey_type.go [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+// https://docs.gitlab.com/ee/api/deploy_keys.html
+// At least 1 of [ProjectId, ProjectIdRef, ProjectIdSelector] required.
+type DeployKeyParameters struct {
+	// The ID or URL-encoded path of the project owned by the authenticated user.
+	// +optional
+	// +immutable
+	// +crossplane:generate:reference:type=github.com/crossplane-contrib/provider-gitlab/apis/projects/v1alpha1.Project
+	// +crossplane:generate:reference:refFieldName=ProjectIdRef
+	// =crossplane:generate:reference:selectorFieldName=ProjectIdSelector
+	ProjectId *string `json:"projectId,omitempty"`
```
[the remark this answers is not in the document]
Comment (lvvorovi) [reply apis/projects/v1alpha1/deploykey_type.go]: Please check if I have fixed it correctly

File: pkg/controller/projects/deploykeys/controller.go [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+	}
+
+	id := meta.GetExternalName(cr)
+
+	if id == "" {
+		return managed.ExternalObservation{}, nil
+	}
+
+	dkArray, _, err := e.client.ListProjectDeployKeys(
+		*cr.Spec.ForProvider.ProjectId,
```
[the remark this answers is not in the document]
Comment (MisterMX) [reply pkg/controller/projects/deploykeys/controller.go]: The referencer will only act if the ref fields are defined. But that is not necessarily the case since ref fields are optional as well. You have to add a custom check.

File: pkg/controller/projects/deploykeys/controller.go [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+	return errors.Wrap(er, errDeleteFail)
+}
+
+func buildReconciler(manager kubeController.Manager, log logging.Logger, name string) *managed.Reconciler {
+	return managed.NewReconciler(manager,
+		resource.ManagedKind(v1alpha1.DeployKeyGroupVersionKind),
+		managed.WithExternalConnecter(buildConnector(manager)),
+		managed.WithInitializers(managed.NewDefaultProviderConfig(manager.GetClient())),
+		managed.WithLogger(log.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(manager.GetEventRecorderFor(name))))
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:225]: Please keep this inside `SetupDeployKey`. It is done so in everywhere and is never used anywhere else again.

File: pkg/controller/projects/deploykeys/controller.go [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+}
+
+func buildConnector(manager kubeController.Manager) *connector {
+	return &connector{kube: manager.GetClient(), newGitlabClientFn: newDeployKeyClient}
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:229]: Same here: Please keep it inside `SetupDeployKey`.

File: pkg/controller/projects/hooks/controller.go
Comment (MisterMX) [inline pkg/controller/projects/hooks/controller.go:1]: These changes here are unrelated to deploy keys. Please create a separate PR for those.

**Review by MisterMX — CHANGES_REQUESTED** [review]

File: apis/projects/v1alpha1/deploykey_type.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (MisterMX) [reply apis/projects/v1alpha1/deploykey_type.go]: Looking good now. Thanks.

File: pkg/controller/projects/deploykeys/controller.go [hunk 22]
```diff
@@ -0,0 +1,274 @@
...
+
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotDeployKey)
+	}
+
+	if cr.Spec.ForProvider.ProjectID == nil {
+		return managed.ExternalObservation{}, errors.New(errProjectIDMissing)
+	}
+
+	ID, err := strconv.Atoi(meta.GetExternalName(cr))
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:98]: Local variables should be lower case.

```suggestion
	id, err := strconv.Atoi(meta.GetExternalName(cr))
```

File: pkg/controller/projects/deploykeys/controller.go [hunk 23]
```diff
@@ -0,0 +1,274 @@
...
+	kubeCoreV1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeTypes "k8s.io/apimachinery/pkg/types"
+	pointer "k8s.io/utils/pointer"
+	kubeController "sigs.k8s.io/controller-runtime"
+	kubeClient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1alpha1 "github.com/crossplane-contrib/provider-gitlab/apis/projects/v1alpha1"
+	clients "github.com/crossplane-contrib/provider-gitlab/pkg/clients"
+	projects "github.com/crossplane-contrib/provider-gitlab/pkg/clients/projects"
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:27]: You only have to provide package aliases if the name is different. Please remove them from if they are not needed.

Please also keep the aliases names consistent across different code files, i.e. `client` is the broadly used name instead of `kubeClient`.

Also: Package aliases should be lowercase.

File: pkg/controller/projects/deploykeys/controller.go [hunk 24]
```diff
@@ -0,0 +1,274 @@
...
+	if err != nil {
+		return errors.Wrap(err, errIDNotAnInt)
+	}
+
+	_, er := e.client.DeleteDeployKey(
+		*cr.Spec.ForProvider.ProjectID,
+		keyID,
+	)
+
+	return errors.Wrap(er, errDeleteFail)
```
Comment (MisterMX) [inline pkg/controller/projects/deploykeys/controller.go:229]: You can reuse `err` here since it is nil anyway. Use `=` instead of `:=`.

```suggestion
	_, err := e.client.DeleteDeployKey(
		*cr.Spec.ForProvider.ProjectID,
		keyID,
	)

	return errors.Wrap(err, errDeleteFail)
```

**Review by MisterMX — CHANGES_REQUESTED** [review]
@lvvorovi This looks pretty good and is close to merge. Some minor remarks to make the code a bit cleaner and more consistent with existing implementations.

**Review by MisterMX — APPROVED** [review]
LGTM. Thanks a lot @lvvorovi!
